backend/push.ts

```diff
@@ -1,74 +1,65 @@
-import { mutators } from "../frontend/mutators";
-import { PushRequest } from "../schemas/push";
-import WebSocket from "ws";
-import { Response } from "schemas/socket";
+import { clients } from "./server";
 
-export async function handlePushRequest(
-  push: PushRequest,
-  docID: string,
-  clientID: string,
-  socket: WebSocket
-) {
+export type PushRequest = {
+  mutations: Mutation[];
+};
+
+export type Client = {
+  clientID: string;
+  pending: Mutation[];
+};
+
+export type Mutation = {
+  id: number;
+  timestamp?: number | undefined;
+};
+
+export type Loop = {
+  run(): Promise<void>;
+};
+
+export type Now = () => number;
+
+export function push(push: PushRequest, client: Client, loop: Loop, now: Now) {
   console.log(
     "Processing push",
     JSON.stringify(push, null, ""),
     "for client",
-    clientID
+    client.clientID
   );
 
-  const t0 = Date.now();
-  await transact(async (executor) => {
-    const tx = new WriteTransactionImpl(executor, clientID, docID);
-
-    let lastMutationID = await getLastMutationID(executor, clientID);
-    console.log("lastMutationID:", lastMutationID);
+  const t0 = performance.now();
 
-    for (let i = 0; i < push.mutations.length; i++) {
-      const mutation = push.mutations[i];
-      const expectedMutationID = lastMutationID + 1;
+  const pending = client.pending;
+  for (const m of push.mutations) {
+    let idx = pending.findIndex((p) => p.id >= m.id);
 
-      if (mutation.id < expectedMutationID) {
-        console.log(
-          `Mutation ${mutation.id} has already been processed - skipping`
-        );
-        continue;
-      }
-      if (mutation.id > expectedMutationID) {
-        console.warn(`Mutation ${mutation.id} is from the future - aborting`);
-        break;
-      }
+    if (idx > -1 && pending[idx].id == m.id) {
+      console.log(`Pending mutation ${m.id} already queued`);
+      continue;
+    }
 
-      console.log("Processing mutation:", JSON.stringify(mutation, null, ""));
+    if (idx == -1) {
+      idx = pending.length;
+    }
 
-      const t1 = Date.now();
-      const mutator = (mutators as any)[mutation.name];
-      if (!mutator) {
-        console.error(`Unknown mutator: ${mutation.name} - skipping`);
-      }
+    // Timestamps must increase monotonically wrt mutation ID.
+    // This only matters while mutations are pending as it is only used
+    // as a heuristic as to the total order to commit in.
+    m.timestamp = clamp(
+      now(),
+      pending[idx - 1]?.timestamp ?? 0,
+      pending[idx]?.timestamp ?? Infinity
+    );
 
-      try {
-        await mutator(tx, mutation.args);
-      } catch (e) {
-        console.error(
-          `Error executing mutator: ${JSON.stringify(mutator)}: ${e}`
-        );
-      }
+    pending.splice(idx, 0, m);
+  }
 
-      lastMutationID = expectedMutationID;
-      console.log("Processed mutation in", Date.now() - t1);
-    }
+  loop.run();
```

Comment:
Might be good/better to separate the actual running of the loop from the signal that the loop should be run. Like here and elsewhere set a flag / fire an event that causes the loop to run in the near future, instead of going into it here. I suggest this for a couple of reasons: first, these functions already do so much, seems like reducing what they do from running the loop to signaling for it might simplify signatures etc. It would decouple the work push() etc have to do from the work that needs to be done as a side effect. Second, pattern wise in an event driven system like this seems more fair / patterny to not synchronously go from one piece of work to another; instead, seems like a call should do its work (eg, push() should enqueue its mutations) and then let others do their work, and if there is follow up work have it scheduled and not run synchronously.